### PR TITLE
Add Basic MVDR module

### PR DESCRIPTION
**Pull request by nateanl**
Summary:
- Add PSD module for computing covariance matrix.
- Add Time-Frequency Masking based MVDR beamforming module.

### CODE REVIEW

**Comment by nateanl** [comment]
I also benchmark the performance of ``torch.einsum`` and ``torch.matmul`` when computing the PSD matrix. Here's the [test script](https://gist.github.com/nateanl/1b519e5e70541ae46334cec493c060fb).

Here is the result:
```
CPU:
[----------- PSD Computation ------------]
                      |  einsum  |  matmul
1 threads: -------------------------------
      [1, 2, 4000.0]  |     43   |    100 
      [1, 2, 8000]    |     62   |    170 
      [1, 2, 16000]   |     97   |    290 
      [1, 2, 32000]   |    400   |    710 
      [1, 4, 4000.0]  |     90   |    240 
      [1, 4, 8000]    |    300   |    600 
      [1, 4, 16000]   |    550   |   1100 
      [1, 4, 32000]   |   2400   |   3400 
      [1, 6, 4000.0]  |    400   |    800 
      [1, 6, 8000]    |    600   |   1200 
      [1, 6, 16000]   |    770   |   2100 
      [1, 6, 32000]   |   2300   |   4200 
      [1, 8, 4000.0]  |    520   |   1000 
      [1, 8, 8000]    |   1300   |   2200 
      [1, 8, 16000]   |   1600   |   3400 
      [1, 8, 32000]   |   9000   |  13000 
4 threads: -------------------------------
      [1, 2, 4000.0]  |     43   |    100 
      [1, 2, 8000]    |     61   |    160 
      [1, 2, 16000]   |    140   |    400 
      [1, 2, 32000]   |    460   |    900 
      [1, 4, 4000.0]  |     90   |    240 
      [1, 4, 8000]    |    290   |    570 
      [1, 4, 16000]   |    550   |   1100 
      [1, 4, 32000]   |   2000   |   3600 
      [1, 6, 4000.0]  |    460   |    700 
      [1, 6, 8000]    |    890   |   1200 
      [1, 6, 16000]   |    700   |   2460 
      [1, 6, 32000]   |   2300   |   4500 
      [1, 8, 4000.0]  |    600   |   1000 
      [1, 8, 8000]    |   1000   |   2500 
      [1, 8, 16000]   |   1600   |   3700 
      [1, 8, 32000]   |   9000   |  13000 
16 threads: ------------------------------
      [1, 2, 4000.0]  |     43   |     99 
      [1, 2, 8000]    |     60   |    160 
      [1, 2, 16000]   |    110   |    330 
      [1, 2, 32000]   |    320   |    720 
      [1, 4, 4000.0]  |     83   |    240 
      [1, 4, 8000]    |    300   |    560 
      [1, 4, 16000]   |    560   |   1500 
      [1, 4, 32000]   |   1800   |   3100 
      [1, 6, 4000.0]  |    330   |    700 
      [1, 6, 8000]    |    700   |   1300 
      [1, 6, 16000]   |    900   |   1800 
      [1, 6, 32000]   |   2100   |   4400 
      [1, 8, 4000.0]  |    510   |   1000 
      [1, 8, 8000]    |   1300   |   2400 
      [1, 8, 16000]   |   1600   |   3600 
      [1, 8, 32000]   |   8900   |  13000 
32 threads: ------------------------------
      [1, 2, 4000.0]  |     40   |     98 
      [1, 2, 8000]    |     60   |    160 
      [1, 2, 16000]   |    110   |    340 
      [1, 2, 32000]   |    320   |    710 
      [1, 4, 4000.0]  |     90   |    240 
      [1, 4, 8000]    |    300   |    600 
      [1, 4, 16000]   |    800   |   1200 
      [1, 4, 32000]   |   2000   |   3200 
      [1, 6, 4000.0]  |    300   |    640 
      [1, 6, 8000]    |    620   |   1300 
      [1, 6, 16000]   |    700   |   1700 
      [1, 6, 32000]   |   2100   |   4400 
      [1, 8, 4000.0]  |    530   |   1000 
      [1, 8, 8000]    |   1400   |   2000 
      [1, 8, 16000]   |   1600   |   3400 
      [1, 8, 32000]   |   9300   |  13000 

Times are in microseconds (us).
```
According to the table, einsum is faster than matmul operation.

**Comment by nateanl** [comment]
PyTorch currently doesn't support batch version of trace method. There's a [linalg.trace PR](https://github.com/pytorch/pytorch/pull/62714) ongoing. We can change it to PyTorch method after the PR is settled.

**Comment by nateanl** [comment]
@Emrys365 Could you help review the MVDR implementation? Thanks!
In terms of the unit tests, do you have any idea to verify the correctness of the solutions?

File: test/torchaudio_unittest/example/beamforming/autograd_test_impl.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        spectrogram = spectrogram.transpose(-2, -3)
+        mask = torch.rand(spectrogram.shape[-3], spectrogram.shape[-1])
+
+        self.assert_grad(transform, [spectrogram, mask])
+
+    @parameterized.expand([
+        param(solution="ref_channel"),
+        param(solution="stv_power"),
+        # evd will fail since the eigenvalues are not distinct
+        # param(solution="stv_evd"),
```
Comment (nateanl) [inline test/torchaudio_unittest/example/beamforming/autograd_test_impl.py:58]: The eigenvalue decomposition solution fails the autograd test. I guess the reason is the eigenvalues are not distinct (i.e. some eigenvalues are close or identical). Is there a way to generate a masked matrix (STFT * mask)  with distinct eigenvalues?

Comment (Emrys365) [reply test/torchaudio_unittest/example/beamforming/autograd_test_impl.py:58]: Maybe following the normal simulation, i.e. simulating a mixture of the white noise and some signal and calculating the idea ratio mask (IRM)?

Comment (nateanl) [reply test/torchaudio_unittest/example/beamforming/autograd_test_impl.py:58]: That's a good idea. I may add an utterance and a room impulse response from open-source  and do the multi-channel signal simulation.

File: examples/beamforming/mvdr.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+    assert input.shape[-1] == input.shape[-2]
+    input = torch.diagonal(input, 0, dim1=-1, dim2=-2)
+    return input.sum(dim=-1)
+
+
+class PSD(torch.nn.Module):
+    r"""Compute cross-channel power spectral density (PSD) matrix.
+
+    Args:
+        normalize (bool): whether normalize the mask over the time dimension
```
Comment (carolineechen) [inline examples/beamforming/mvdr.py:50]: nit: `(bool, optional)`

the torchaudio convention for docstrings uses `optional` to refer there being a default value (and not the `Optional` type). For instance, `(T, optional)` if for a variable of type T with a default value, and `(T or None, optional)` for a variable of type `Optional[T]` with a default value. I see some other inconsistencies with this in this PR

File: examples/beamforming/mvdr.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        assert solution in ["ref_channel", "stv_evd", "stv_power"],\
+            "Unknown solution provided. Must be one of [``ref_channel``, ``stv_evd``, ``stv_power``]."
+        self.ref_channel = ref_channel
+        self.solution = solution
+        self.diag_loading = diag_loading
+        self.diag_eps = diag_eps
+        self.online = online
+        self.psd = torch.jit.trace(PSD(), (torch.rand(2, 129, 6, 100, dtype=torch.cdouble), torch.rand(2, 129, 100)))
+
+        psd_s: Optional[torch.Tensor] = torch.zeros(1)
```
Comment (carolineechen) [inline examples/beamforming/mvdr.py:122]: Is there a reason why these are of type `Optional`? I don't see a scenario where it equals to `None`, and it's typed as `torch.Tensor` in function inputs

File: examples/beamforming/mvdr.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        """
+        if self.psd_s == 0:
+            self.psd_s = psd_s
+            self.psd_n = psd_n
+            self.mask_sum_s = mask_s.sum(dim=-1)
+            self.mask_sum_n = mask_n.sum(dim=-1)
+            return self.get_mvdr_vector(psd_s, psd_n, reference_vector, solution, diagonal_loading, diag_eps, eps)
+        else:
+
+            psd_s = self.update_psd_speech(psd_s, mask_s)
```
Comment (carolineechen) [inline examples/beamforming/mvdr.py:170]: thoughts on using a function name like `get_updated_psd_speech` or having the function directly modify `self.psd_s`? The current function name makes it seem as if updating is part of the function

**Review by carolineechen — COMMENTED** [review]
A few minor comments:

**Review by Emrys365 — COMMENTED** [review]
The MVDR implementation looks good to me. I just made some minor comments.

File: examples/beamforming/mvdr.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+            reference_vector (torch.Tensor): one-hot reference channel matrix
+
+        Returns:
+            torch.Tensor: the enhanced STFT
+                Tensor of dimension (..., freq, channel, 1)
+        """
+        phi = torch.linalg.solve(psd_n, psd_s)  # psd_n.inv() @ stv
+        stv = torch.einsum("...fec,...c->...fe", [phi, reference_vector])
+        stv = stv.unsqueeze(-1)
+        for _ in range(1):
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/beamforming/mvdr.py:295]: I followed the code in ESPNet and forgot to refactor it😅. https://github.com/espnet/espnet/blob/master/espnet2/enh/layers/beamformer.py#L54

Do you think one iteration is good enough?

[the remark this answers is not in the document]
Comment (Emrys365) [reply examples/beamforming/mvdr.py:295]: I think 3 iterations might be fine. Sometimes only 1 iteration may be inaccurate.

**Comment by nateanl** [comment]
I will keep the "complex-valued mask support" as a future plan since I'm not sure of the formula to normalize the psd matrix by the mask. Here is my proposed method:
```
S = X * mask.unsqueeze(-3)  # complex multiplication
psd = torch.einsum("...cft,...eft->...ftce", [S, S.conj()])
if self.normalize:
    mask = mask.abs() * mask.abs()  # since the numerator multiplies the mask twice
    mask = mask.sum(dim=-1, keepdim=True)
    psd = psd / (mask + eps)
```

File: examples/beamforming/mvdr.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+        Returns:
+            torch.Tensor: the enhanced STFT
+                Tensor of dimension (..., freq, channel, 1)
+        """
+        w, v = torch.linalg.eig(psd_s)  # (..., freq, channel, channel)
+        _, indices = torch.max(w.abs(), dim=-1, keepdim=True)
+        indices = indices.unsqueeze(-1)
+        stv = v.gather(-1, indices.expand(psd_s.shape[:-1] + (1,)))  # (..., freq, channel, 1)
+        return stv
```
[the remark this answers is not in the document]
Comment (popcornell) [reply examples/beamforming/mvdr.py]: @nateanl, sorry for my late response. I ve tried several formulations this spring and this one [1] was the most robust when training e2e.

[1] Souden, M., Benesty, J., & Affes, S. (2009). On optimal frequency-domain multichannel
linear filtering for noise reduction. IEEE Transactions on audio, speech, and language processing, 18(2), 260-276.

[the remark this answers is not in the document]
Comment (nateanl) [reply examples/beamforming/mvdr.py]: Thanks Samuele! The formula refers to the reference channel selection solution in the module, right? 
 ```w = \Phi_NN^-1 \Phi_SS u / trace( \Phi_NN^-1 \Phi_SS)```
I will benchmark the performances of all solutions and compare with yours. They should be consistent.

[the remark this answers is not in the document]
Comment (popcornell) [reply examples/beamforming/mvdr.py]: Yep looks like the formula you already implemented as `ref_channel`
